### PR TITLE
Issue 280 - Refactor cell content

### LINKED ISSUE #280: Refactor cell content in preparation for data types
Currently the cell content is a <CellInput /> that contains all possible implementations baked into one (label, input, dropdown). Isolate these roles from one another and move the logic to choose which one to use into the derived/cell/contents calculation.

Prep work for #166

### PR DESCRIPTION
**Pull request by Marc-Andre-Rivet**
Closes #280 in preparation for #166.

Breaks up the existing CellInput component into all its flavors (label, input, dropdown) and moving the choice into the derived/cell/contents calculation. Separating the logic from the flavors implementation details will make it easier to support the additional data types and scenarios that pop up in the data types epic.

This is all prep work for the data types. There's no additional feature here -- additional sanity may be needed but the tests cover a pretty broad range of cases already.

Not rebasing as there are not that many changes despite the large amount of commits (this branch is a child of the virtualization branch -- that's why!)

### CODE REVIEW

File: src/dash-table/components/CellDropdown/index.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+interface IProps {
+    active: boolean;
+    clearable?: boolean;
+    dropdown: IDropdownOptions;
+    onChange: (e: ChangeEvent) => void;
+    value: any;
+}
+
+export default class CellDropdown extends PureComponent<IProps> {
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/components/CellDropdown/index.tsx]: Standalone Dropdown component pulled out of pre-existing CellInput comp

File: src/dash-table/components/CellInput/index.tsx [hunk 2]
```diff
@@ -1,30 +1,33 @@
...
+
+interface ICellState {
+    value: any;
+}
 
-    public static defaultProps: ICellDefaultProps = {
-        conditionalDropdowns: [],
-        type: ColumnType.Text
-    };
+export default class CellInput extends PureComponent<ICellProps, ICellState> {
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/components/CellInput/index.tsx]: Standalone Input component -- what's left after removing Label and Dropdown from the existing implementation

File: src/dash-table/components/CellLabel/index.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+} from 'react';
+
+interface IProps {
+    className: string;
+    onClick?: (e: MouseEvent) => void;
+    onDoubleClick?: (e: MouseEvent) => void;
+    value: any;
+}
+
+export default class CellLabel extends PureComponent<IProps> {
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/components/CellLabel/index.tsx]: Standalone Label component pulled out of pre-existing CellInput comp -- as before we use this to display non-editable fields, fallback dropdown fields and to optimize inputs while cell is not active.

File: src/dash-table/derived/cell/contents.tsx [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    switch (type) {
+        case ColumnType.Text:
+        case ColumnType.Numeric:
+            return (!active || !editable) ? CellType.Label : CellType.Input;
+        case ColumnType.Dropdown:
+            return (!dropdown || !editable) ? CellType.Label : CellType.Dropdown;
+        default:
+            return CellType.Label;
+    }
+}
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/derived/cell/contents.tsx]: What type of cell content

File: src/dash-table/derived/cell/contents.tsx [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+            const isEditable = isCellEditable(editable, column.editable);
+
+            const className = [
+                ...(active ? ['input-active'] : []),
+                ...(isFocused ? ['focused'] : ['unfocused']),
+                ...['dash-cell-value']
+            ].join(' ');
+
+            switch (getCellType(active, isEditable, dropdown, column.type)) {
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/derived/cell/contents.tsx]: Once the type of cell content is decided, apply the right one

File: src/dash-table/components/CellDropdown/index.tsx (hunk 1 above)
Comment (Marc-Andre-Rivet) [reply src/dash-table/components/CellDropdown/index.tsx]: Well it still uses react-select in the end :(

File: src/dash-table/derived/cell/contents.tsx [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    editable: boolean,
+    dropdown: any,
+    type: ColumnType = ColumnType.Text
+) {
+    switch (type) {
+        case ColumnType.Text:
+        case ColumnType.Numeric:
+            return (!active || !editable) ? CellType.Label : CellType.Input;
+        case ColumnType.Dropdown:
+            return (!dropdown || !editable) ? CellType.Label : CellType.Dropdown;
```
[the remark this answers is not in the document]
Comment (Marc-Andre-Rivet) [reply src/dash-table/derived/cell/contents.tsx]: Good point. I'll see if I can tighten the typing here

File: src/dash-table/components/CellDropdown/index.tsx [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    private setFocus() {
+        const { active } = this.props;
+        if (!active) {
+            return;
+        }
+
+        const dropdown = this.refs.dropdown as any;
+
+        if (dropdown && document.activeElement !== dropdown) {
+            // Limitation. If React >= 16 --> Use React.createRef instead to pass parent ref to child
```
[the remark this answers is not in the document]
Comment (Marc-Andre-Rivet) [reply src/dash-table/components/CellDropdown/index.tsx]: This is what I have in mind: ref forwarding (added in React 16.3)
https://reactjs.org/docs/forwarding-refs.html

[the remark this answers is not in the document]
Comment (valentijnnieman) [reply src/dash-table/components/CellDropdown/index.tsx]: Ah gotcha. Yeah that would be nice!

File: src/dash-table/derived/cell/contents.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (Marc-Andre-Rivet) [reply src/dash-table/derived/cell/contents.tsx]: Improved typing so we know we have either dropdown values (array) or undefined -- there were two sets of independent interfaces used to define dropdown labels and values -- removed one, kept the other. Lots of small additional changes all over the place for that consolidation.

**Pull request merged**